presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterThroughWindow.java

```diff
@@ -0,0 +1,190 @@
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.window;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterThroughWindow
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter().with(source().matching(window()));
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+    private final FunctionId rowNumberFunctionId;
+
+    public PushdownFilterThroughWindow(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+        PlanNodeIdAllocator idAllocator = context.getIdAllocator();
+
+        PlanNode source = getOnlyElement(lookup.resolveGroup(node.getSource()).collect(Collectors.toList()));
+
+        TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
+
+        if (source instanceof WindowNode && isRowNumber((WindowNode) source, rowNumberFunctionId) && isOptimizeTopNRowNumber(session)) {
+            WindowNode windowNode = (WindowNode) source;
+            Symbol rowNumberSymbol = getOnlyElement(windowNode.getWindowFunctions().entrySet()).getKey();
+            OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+            if (upperBound.isPresent()) {
+                if (upperBound.getAsInt() <= 0) {
+                    return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+                }
+                if (windowNode.getOrderingScheme().isPresent()) {
+                    source = new TopNRowNumberNode(idAllocator.getNextId(),
+                            windowNode.getSource(),
+                            windowNode.getSpecification(),
+                            getOnlyElement(windowNode.getWindowFunctions().keySet()),
+                            upperBound.getAsInt(),
+                            false,
+                            Optional.empty());
+                    return Result.ofPlanNode(rewriteFilterSource(node, source, rowNumberSymbol, upperBound.getAsInt(), session, types));
+                }
+            }
+        }
+        return Result.empty();
+    }
+
+    private PlanNode rewriteFilterSource(FilterNode filterNode, PlanNode source, Symbol rowNumberSymbol, int upperBound, Session session, TypeProvider types)
```

Comment:
This is missing the same treatment as the other rule.

Reply:
Oh, sorry, I'll fix that.